internal/pkg/service/common/etcdop/consumer.go

```diff
@@ -0,0 +1,142 @@
+type WatchConsumer[E any] struct {
+	logger      log.Logger
+	stream      <-chan WatchResponseE[E]
+	forEachFn   func(events []E, header *Header, restart bool)
+	onCreated   onWatcherCreated
+	onRestarted onWatcherRestarted
+	onError     onWatcherError
+}
+
+type WatchStreamE[E any] <-chan WatchResponseE[E]
+
+type (
+	onWatcherCreated   func(header *Header)
+	onWatcherRestarted func(reason string, delay time.Duration)
+	onWatcherError     func(err error)
+)
+
+func newConsumer[E any](logger log.Logger, stream <-chan WatchResponseE[E]) WatchConsumer[E] {
+	return WatchConsumer[E]{
+		logger: logger,
+		stream: stream,
+	}
+}
+
+func (s WatchStreamE[E]) SetupConsumer(logger log.Logger) WatchConsumer[E] {
+	return newConsumer[E](logger, s)
+}
+
+func (o WatchConsumer[E]) WithForEach(v func(events []E, header *Header, restart bool)) WatchConsumer[E] {
+	o.forEachFn = v
+	return o
+}
+
+func (o WatchConsumer[E]) WithOnCreated(v onWatcherCreated) WatchConsumer[E] {
+	o.onCreated = v
+	return o
+}
+
+func (o WatchConsumer[E]) WithOnRestarted(v onWatcherRestarted) WatchConsumer[E] {
+	o.onRestarted = v
+	return o
+}
+
+func (o WatchConsumer[E]) WithOnError(v onWatcherError) WatchConsumer[E] {
+	o.onError = v
+	return o
+}
+
+func (o WatchConsumer[E]) StartConsumer(wg *sync.WaitGroup) (initErr <-chan error) {
+	initErrCh := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		// The flag restart=true is send with the first events batch after the "restarted" event, see WatchConsumer.forEachFn.
+		restart := false
+
+		// See watchErrorThreshold
+		var lastErrorAt time.Time
+
+		// Channel is closed when the watcher context is cancelled,
+		// so the context does not have to be checked here.
+		for resp := range o.stream {
+			switch {
+			case resp.InitErr != nil:
+				// Initialization error, the channel will be closed in the beginning of the next iteration.
+				// Signal the problem via InitErr channel.
+				// It is fatal error (e.g., no network connection), the app should be stopped and restarted.
+				initErrCh <- resp.InitErr
+				close(initErrCh)
+			case resp.Err != nil:
+				// An error occurred, it is logged.
+				// If it is a fatal error, then it is followed
+				// by the "Restarted" event handled bellow,
+				// and the operation starts from the beginning.
+				//
+				// ErrCompacted or ErrLeaderChanged occurs even during cluster normal operation,
+				// so the error is logged with warning log level.
+				//
+				// It is suspicious if a short time has passed between two errors,
+				// then the error is logged with error log level.
+				if interval := time.Since(lastErrorAt); interval > watchErrorThreshold {
+					o.logger.Warn(resp.Err)
+				} else {
+					o.logger.Error(errors.Errorf(`%w (previous error %s ago)`, resp.Err, interval))
+				}
+				lastErrorAt = time.Now()
+				if o.onError != nil {
+					o.onError(resp.Err)
+				}
```

Comment:
Tu je zmena, chybu logujeme iba ako wargning (poriesi to restart).
Iba ked sa vyskytnu 2 chyby za `5s`, tak to logujeme ako error.